libsolidity/analysis/DeclarationTypeChecker.cpp

```diff
@@ -309,6 +309,12 @@ void DeclarationTypeChecker::endVisit(ArrayTypeName const& _typeName)
 			m_errorReporter.typeError(3208_error, length->location(), "Array with fractional length specified.");
 		else if (*lengthValue < 0)
 			m_errorReporter.typeError(3658_error, length->location(), "Array with negative length specified.");
+		else if (lengthValue->numerator() > TypeProvider::uint256()->max())
+			m_errorReporter.typeError(
+				1847_error,
+				length->location(),
+				"Array length too large, maximum is 2^256-1."
+			);
```

Comment:
Please add a sanity check here:
```suggestion
		else
		{
			solAssert(lengthValue->denominator() == 1);
			if (lengthValue->numerator() > TypeProvider::uint256()->max())
				m_errorReporter.typeError(
					1847_error,
					length->location(),
					"Array length too large, maximum is 2^256-1."
				);
		}
```
and also include a short comment explaining why the comparison is being done this way (this is not something obvious).

Reply:
Reverting to the comparison so we don't need this assertion or explanation.